src/runtime_src/doc/toc/xrt_native_apis.rst

```diff
@@ -0,0 +1,618 @@
+              
+           xrtKernelClose(kernel);
+
+
+In the above code block
+
+              - The compute unit named "foo_1" (name syntax: "kernel_name:{cu_name}") is opened exclusively.
+              - The Register Read/Write operation is performed. 
+              - Closed the kernel
+              
+**C++**: The equivalent C++ API example
+
+.. code:: c
+      :number-lines: 35
+       
+           int read_data; 
+           int write_data = 7; 
+              
+           auto krnl = xrt::kernel(device, "foo:{foo_1}", xclbin_uuid, true); 
+
+           read_data = kernel.read_register(READ_OFFSET);
+           kernel.write_register(WRITE_OFFSET,write_data); 
+              
+
+     
+Executing the kernel
+~~~~~~~~~~~~~~~~~~~~
+
+Execution of the kernel is associated with a **Run** handle (or object). The kernel can be executed by the API ``xrtKernelRun`` (in C++ overloaded operator ``xrt::kernel::()``) that takes all the kernel arguments in order. The kernel execution API returns a run handle (or object) corresponding to the execution. 
+
+
+.. code:: c
+      :number-lines: 50
+       
+           // 1st kernel execution
+           xrtRunHandle run = xrtKernelRun(kernel, buf_a, buf_b,  scalar_1); 
+           xrtRunWait(run);
+    
+           // 2nd kernel execution with just changing 3rd argument
+           xrtRunSetArg(run,2,scalar_2); // Arguments are specified starting from 0
+           xrtRunStart(run);
+           xrtRunWait(run);
+
+           // Close the run handle
+           xrtRunClose(run);
+
+Note the following APIs regarding  the above example
+
+   - The kernel is executed by ``xrtKernelRun`` API by specifying all its arguments to obtain a Run handle
+   - The API ``xrtKernelRun`` is non-blocking. It returns as soon as it submits the job without waiting for the kernel's actual execution start.  
+   - The host code uses ``xrtRunWait`` API to block the current thread and wait till the kernel execution is finished.       
+   - After a run is finished, the same run handle can be reused to execute the kernel multiple times if desired. 
+     
+       - API ``xrtRunSetArg`` is used to set one or more arguments, in the example above only the last (3rd) argument is changed before the second execution
+       - API ``xrtRunStart`` is used to execute the kernel using the run handle. 
+   - API ``xrtRunClose`` is used to close the Run handle.  
+ 
+   
+**C++**: The equivalent C++ code
+
+In C++ the ``xrt::kernel`` class provides **overloaded operator ()** to execute the kernel with a comma-separated list of arguments.  
+
+.. code:: c++
+      :number-lines: 50
+   
+           // 1st kernel execution
+           auto run = kernel(buf_a, buf_b, scalar_1); 
+           run.wait();
+    
+           // 2nd kernel execution with just changing 3rd argument
+           run.set_arg(2,scalar_2); // Arguments are specified starting from 0 
+           run.start();
+           run.wait();
```

Comment:
Strange API but why not.
`run` could be named `executor` or `execution`? It is strange to have a verb as a class.

Reply:
What is the life-time? RAII?

Reply:
Why are you concerned about lifetime?  This is C++ of course it is managed.

Reply:
nothing needed to be done for this PR

Reply:
> Why are you concerned about lifetime? This is C++ of course it is managed.

I review a lot of Xilinx C++ code, so I have some reasons to be naturally worried about a spam of pointers, `malloc`/`free`/`new`/`delete`/... :-( and I want to double check... ;-)
Where is the basic C++ behavior of the API is explained, the equivalent of Section "3.8.12 Managing object lifetimes" on page 47, Section "4.5.3 Common reference semantics" on page 60 or Section "4.5.4 Common by-value semantics" on page 62 of 
https://www.khronos.org/registry/SYCL/specs/sycl-2020-provisional.pdf ?

Reply:
Yes, my point is that if there is anything that requires { dosomething(); cleanup(); } then that is a bug.  Never ever should clients of the C++ APIs have to clean up, it would be a messed up API if they had to.  May still be messed up, but at least no resources should be leaked ever.